### ISSUE TITLE
terrain loading error

### ISSUE THREAD

**Issue by jbo023** [AnalyticalGraphicsInc/cesium#4894]
Ho, 

we just found a weird bug in the terrain loading code. Sometimes in the front of the screen the terrain is not loaded (see screenshot). Unfortunately this Bug does not occur all the time. If you disable the webbrowser file cache it occurs more often. 

I narrowed it down to the following Pull Request https://github.com/AnalyticalGraphicsInc/cesium/pull/4616. Before the merge I could not reproduce the problem. 

It is also interesting that for the same startview, a refresh of the page does not always request the same amount of terrain tiles. As an example for another application I counted between 130 and 200 terrain requests.

![image](https://cloud.githubusercontent.com/assets/4146115/22157609/a53d2db8-df38-11e6-86c0-f377169e0a8b.png)

```
var viewer = new Cesium.Viewer('cesiumContainer');

var cesiumTerrainProviderMeshes = new Cesium.CesiumTerrainProvider({
    url : 'http://virtualcitymaps.s3-website.eu-central-1.amazonaws.com/terrain/',
    requestWaterMask : false,
    requestVertexNormals : true
});
viewer.terrainProvider = cesiumTerrainProviderMeshes;


var cam = viewer.scene.camera;

var cameraPosition = new Cesium.Cartesian3(3950334.446443219, 769922.5676672924, 4932162.0991373975);
var cam = viewer.scene.camera;
var cameraOptions = {
"heading": 0.05446154055800534,
"pitch":-0.6709337075080422,
"roll": 0.00022910782255536333
};

var direction = new Cesium.Cartesian3(-0.9887510456113379, -0.1492704873805369, 0.009470554348256777);
var up = new Cesium.Cartesian3(0.0042078055935553796,0.03553292930862533,0.9993596476278372);


cam.setView({
"destination": cameraPosition,
"orientation": {
"direction":direction,
"up":up
}
});
```

**Comment by jbo023** [comment]
Terrain is loaded correctly (197 requests, browser cache active)
![image](https://cloud.githubusercontent.com/assets/4146115/22158149/32d04924-df3b-11e6-96f7-f138e320cab4.png)

Terrain is loaded incorrectly (144 requests, browser cache disabled)
![image](https://cloud.githubusercontent.com/assets/4146115/22158170/4a2fcb9e-df3b-11e6-9206-d8565797d627.png)


You can see that in the the second image the terrain loading algorithm starts with lower levels of the terrain, while in the correct version the algorithm starts with the higher terrain levels.

**Comment by pjcozzi** [comment]
@jbo023 thanks for the detailed report and narrowing this down.

@kring any ideas?

**Comment by kring** [comment]
Are you able to reproduce this with STK World Terrain, too, or only with your custom terrain tileset?  Is your tileset quantized-mesh or heightmap?

**Comment by jbo023** [comment]
It is a custom quantized-mesh terrain tileset. The link in the sandcastle has to be changed to this terrain.
http://d35ei6ur3bjvr1.cloudfront.net/thueringen/terrain_5/
I will also check if i can reproduce this using the STK World Terrain.

**Comment by jbo023** [comment]
Hi, i could also reproduce the problem with the STK World Terrain. 

Also I noticed that the number of terrain request for the same view are more or less random.
The bug seems to be related to the speed the tiles will be delivered.

**Comment by kring** [comment]
> Hi, i could also reproduce the problem with the STK World Terrain.

I can reproduce it using your code as well.  Thanks!  I'll take a look soon.

> Also I noticed that the number of terrain request for the same view are more or less random.

Sure, that's to be expected.  It's certainly non-deterministic, depending on how long it takes for tile requests to come back and in what order.  To fix this, and make Cesium load exactly the tiles needed for a particular view, we would need to make per-tile min/max heights available without downloading tiles.

**Comment by kring** [comment]
@jbo023 a fix is in #4969, let me know if it works for you.

**Comment by hpinkos** [comment]
I run into this all of the time when I have a camera flight go close to the ground at the start of an application:

``` js
var viewer = new Cesium.Viewer('cesiumContainer');
viewer.terrainProvider = new Cesium.CesiumTerrainProvider({
    url : 'https://assets.agi.com/stk-terrain/v1/tilesets/world/tiles',
    requestWaterMask : true,
    requestVertexNormals : true
});
var center = new Cesium.Cartesian3(960309.3317601935, -4954000.560677782, 3888228.4347712533);
var bs = new Cesium.BoundingSphere(center, 164.94449695026205);
viewer.scene.camera.flyToBoundingSphere(bs, {offset: new Cesium.HeadingPitchRange(0.0, -0.5, bs.radius * 2.0)});
```

In this case, the tiles decided never to load in:

![terrain](https://user-images.githubusercontent.com/3451886/33955930-330f0fb6-e00b-11e7-9f27-55710900f645.gif)

**Comment by hpinkos** [comment]
It looks like #4969 fixes the problems I'm seeing, although I still see that flickering for a few seconds before the tiles finally start to load in

**Comment by hpinkos** [comment]
@jbo023 this issue has been fixed and will be included in the upcoming 1.42 release, available on Feb 1st. Thanks@
